stat/spatial/spatial.go

```diff
@@ -0,0 +1,131 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package spatial // import "gonum.org/v1/gonum/stat/spatial"
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/mat"
+	"gonum.org/v1/gonum/stat"
+)
+
+// TODO(kortschak): Implement weighted routines.
+// TODO(kortschak): Make use of banded matrices when they exist in mat.
+
+// GetisOrdGStar returns the Local Getis-Ord G*i statistic for element of of the
+// weighted data using the provided locality matrix. The returned value is a z-score.
+//
+//  G^*_i = num_i / den_i
+//
+//  num_i = \sum_j (w_{ij} x_j) - \bar X \sum_j w_{ij}
+//  den_i = S \sqrt((n \sum_j w_{ij}^2 - (\sum_j w_{ij})^2)/(n - 1))
+//  \bar X = (\sum_j x_j) / n
+//  S = \sqrt((\sum_j x_j^2)/n - (\bar X)^2)
+//
+// GetisOrdGStar will panic if locality is not a square matrix with dimensions the
+// same as the length of data or if i is not a valid index into data.
+//
+// See doi.org/10.1111%2Fj.1538-4632.1995.tb00912.x.
+//
+// Weighted Getis-Ord G*i is not currently implemented and GetisOrdGStar will
+// panic if weights is not nil.
+func GetisOrdGStar(i int, data, weights []float64, locality mat.Matrix) float64 {
+	if weights != nil {
+		panic("spatial: weighted data not yet implemented")
+	}
+	r, c := locality.Dims()
+	if r != len(data) || c != len(data) {
+		panic("spatial: data length mismatch")
+	}
+
+	n := float64(len(data))
+	mean := stat.Mean(data, weights)
+	var ss, dwd, dww, sw float64
+	for j, v := range data {
+		ss += v * v
+		w := locality.At(i, j)
+		sw += w
+		dwd += w * v
+		dww += w * w
+	}
+	s := math.Sqrt(ss/n - mean*mean)
+
+	return (dwd - mean*sw) / (s * math.Sqrt((n*dww-sw*sw)/(n-1)))
+}
+
+// GlobalMoransI performs Global Moran's I calculation of spatial autocorrelation
+// for the given data using the provided locality matrix. GlobalMoransI returns
+// Moran's I, Var(I) and the z-score associated with those values.
+// GlobalMoransI will panic if locality is not a square matrix with dimensions the
+// same as the length of data.
+//
+// See https://doi.org/10.1111%2Fj.1538-4632.2007.00708.x.
+//
+// Weighted Global Moran's I is not currently implemented and GlobalMoransI will
+// panic if weights is not nil.
+func GlobalMoransI(data, weights []float64, locality mat.Matrix) (i, v, z float64) {
+	if weights != nil {
+		panic("spatial: weighted data not yet implemented")
+	}
+	if r, c := locality.Dims(); r != len(data) || c != len(data) {
+		panic("spatial: data length mismatch")
+	}
+	mean := stat.Mean(data, nil)
+
+	// Calculate Moran's I for the data.
+	var num, den float64
+	for i, xi := range data {
+		zi := xi - mean
+		den += zi * zi
+		for j, xj := range data {
+			zj := xj - mean
+			num += locality.At(i, j) * zi * zj
+		}
+	}
+	i = (float64(len(data)) * num) / (mat.Sum(locality) * den)
```

Comment:
Sorry, the question was about "I'd do (N/W)*(num/den)". Do you want that change for readability?